src/articles/FindArticles.js

```diff
@@ -7,9 +7,29 @@ import * as s from "./FindArticles.sc";
 import LoadingAnimation from "../components/LoadingAnimation";
 import { setTitle } from "../assorted/setTitle";
 import strings from "../i18n/definitions";
+import Reminder from "./Reminder";
+import ExtensionMessage from "../components/ExtensionMessage";
+import Feature from "../features/Feature";
+import LocalStorage from "../assorted/LocalStorage";
+import { runningInChromeDesktop } from "../utils/misc/browserDetection";
+import { checkExtensionInstalled } from "../utils/misc/extensionCommunication";
 
 export default function NewArticles({ api }) {
   const [articleList, setArticleList] = useState(null);
+  const [hasExtension, setHasExtension] = useState(false);
+  const [extensionMessageOpen, setExtensionMessageOpen] = useState(false);
+  const [displayedExtensionPopup, setDisplayedExtensionPopup] = useState(false);
+
+  useEffect(() => {
+    setDisplayedExtensionPopup(LocalStorage.displayedExtensionPopup());
+    if (runningInChromeDesktop() && Feature.extension_experiment1() && !displayedExtensionPopup) {
+        checkExtensionInstalled(setHasExtension);
+        if(!hasExtension){
```

Comment:
THANKS FOR MOVING THE `checkExtensionInstalled` to its own file. 
The FindArticles looks much better now!

I think this has to be taken out of the `useEffect` maybe?

```js
  useEffect(() => {
    setDisplayedExtensionPopup(LocalStorage.displayedExtensionPopup());
    if (
      runningInChromeDesktop() &&
      Feature.extension_experiment1() &&
      !displayedExtensionPopup
    ) {
      checkExtensionInstalled(setHasExtension);
    }
  }, []);

  if (!hasExtension) {
    setExtensionMessageOpen(true);
  }

```

Does it work that way? 

It seems to me that with the current approach, it could be possible 
that we have a race condition, in which line 27 gets executed before
the callback inside `checkExtensionIstalled` is called; and we thus,
don't have the correct picture of the state of the installed extension.
And given that the conditional on line 27 is verified inside of the useEffect
with second param [], it's never going to be reevaluated.

If we take it out, then it is going to be checked, whenever the 
`checkExtensionIstalled` changes the state of the setHasExtension.

Do you see what I mean? 

Finally, for the situation in which the 
extension is installed, but the chrome messaging is slow, it 
would probably make sense to start with hasExtension = true as 
a default state? such that we don't display the message, then 
get the answer that the extension is present, and we hide it, 
and the user might see a flicker?

Reply:
I updated the code now. I still had to put "(!hasExtension)" inside of an useEffect - otherwise it created an infinite loop. But I added it to another one - and only run that if hasExtension updates. I also put hasExtension = true. And then I added another return inside of Reminder - to display to the users, who does not use a chrome browser.